### DISCUSSION TITLE
Having Problems With `ses-ava`

### DISCUSSION THREAD

**Discussion by anilhelvaci** [endojs/endo#1595] [Q&A] [answered]
## Problem Description

I've been trying to write some `swingsetTests` in my project to upgrade a smart contract as it is in [test-coveredCall-service-upgrade.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/test/swingsetTests/upgradeCoveredCall/test-coveredCall-service-upgrade.js). But I'm having problems with my dummy test below:

```js
import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';

test('initial', async t => {
  t.pass('This is a dummy test');
})
```

Here's the error message I get:

```
Uncaught exception in test/swingsetTests/test-crabble-upgrade.js

  AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: The expression evaluated to a falsy value:

    assert(refs.runnerChain)

    at async Promise.all (index 0)

SES_UNHANDLED_REJECTION: (AssertionError#1)
  ✘ test/swingsetTests/test-crabble-upgrade.js exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```

When I copy/paste the contents of [prepare-test-env-ava.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/SwingSet/tools/prepare-test-env-ava.js) to my test file directly like below I get a different error message;

Test File

```js
import '@endo/init/pre-bundle-source.js';

import '@agoric/swingset-vat/tools/prepare-test-env.js';

import '@endo/ses-ava/exported.js';

import { wrapTest } from '@endo/ses-ava';
import rawTest from 'ava';

export const test = wrapTest(rawTest);

export const VatData = globalThis.VatData;
assert(VatData);

test('initial', async t => {
  t.pass('This is a dummy test');
})
```

Error Message

```
Uncaught exception in test/swingsetTests/test-crabble-upgrade.js

  TypeError: Cannot redefine property: toString
    at defineProperty (<anonymous>)
    at async Promise.all (index 0)

  › defineProperty (<anonymous>)

  ✘ test/swingsetTests/test-crabble-upgrade.js exited with a non-zero exit code: 1
  ─

  1 uncaught exception
```

When I comment out the line that imports `wrapTest`, the test passes;

Test file `wrapTest` removed

```js
import '@agoric/swingset-vat/tools/prepare-test-env.js';

import '@endo/ses-ava/exported.js';

// import { wrapTest } from '@endo/ses-ava';
import rawTest from 'ava';

export const test = rawTest;

export const VatData = globalThis.VatData;
assert(VatData);

test('initial', async t => {
  t.pass('This is a dummy test');
})
```

Above test passes.

I'm able to make [test-coveredCall-service-upgrade.js](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/test/swingsetTests/upgradeCoveredCall/test-coveredCall-service-upgrade.js) work so I'm guessing the issue is related to my project setup. To replicate the `zoe` setup I copy/pasted all (dev)dependencies from [zoe/package.json](https://github.com/Agoric/agoric-sdk/blob/master/packages/zoe/package.json) and results were still negative. I would appreciate any ideas about what might be the cause 🙏 Until this question is resolved I'll keep on my development using the piece of code below as it works:

```js
import '@agoric/swingset-vat/tools/prepare-test-env.js';
import test from 'ava';
```

cc @dckc 

## Possibly Related Issues/PRs
#1465 [#5774](https://github.com/Agoric/agoric-sdk/pull/5774) #1235

**Comment by dckc** [comment]
Hi @anilhelvaci; thanks for sharing your experience... it's not obvious to me what's going on.

@kriskowal here's hoping for a moment to ask you about this

**Comment by kriskowal** [comment]
I’m equally perplexed but also not surprised. The Cannot redefine toString suggests that the object in question is frozen, but which object is a mystery. It doesn’t look to me like `ses-ava` modifies any shared object, and does take care to use `defineProperty` in case the underlying object is frozen. That suggests that, even if there’s more than one instance of `ses-ava` in your program (since `import "@endo/ses-ava"` can resolve to different physical locations depending on the package it’s imported from), there should not be a problem.

I agree your isolation suggests that there is a problem in `ses-ava`, but the `assert(refs.runnerChain)` seems unrelated.

Please let me know if you find a useful clue in my wild speculation.

**Comment by anilhelvaci** [comment, marked as answer]
For those who run into this issue; this turned out to be the cause for me => https://github.com/avajs/ava/issues/2946#issuecomment-1712945678
